Added support for byval variadic arguments

ccadar opened the pull request.

ccadar commented:
Never say never -- this solves a bug that is almost 11 years old, reported first in July 2009! (see #189)

kren1 commented on lib/Core/Executor.cpp:
```diff
@@ -1356,15 +1361,13 @@ static inline const llvm::fltSemantics *fpWidthToSemantics(unsigned width) {
-                           KInstruction *ki,
-                           Function *f,
-                           std::vector< ref<Expr> > &arguments) {
+void Executor::executeCall(ExecutionState &state, KInstruction *ki, Function *f,
```
This format isn't consistent with the rest of Executor

kren1 commented on lib/Core/Executor.cpp:
```diff
@@ -1396,28 +1399,30 @@ void Executor::executeCall(ExecutionState &state,
-                               AddExpr::create(arguments[0], 
-                                               ConstantExpr::create(16, 64)),
-                               ConstantExpr::create(0, 64), 0); // reg_save_area
+        executeMemoryOperation(
```
I find the previous format clearer.

kren1 commented on test/Feature/VarArgAlignment.c:
```diff
@@ -0,0 +1,88 @@
+void test1(int x, ...) {
+  va_list ap;
+  va_start(ap, x);
+  int i1 = va_arg(ap, int);
```
Wouldn't it be nice to test `char` and `short` too?

ccadar replied:
You are right, although in this case I'd be hesitant to change it, as I played with these tests to check various alignments.   But we could add another test if you think it's necessary.

ccadar replied:
Actually, I was just reminded by the compiler that you cannot pass chars and shorts to a variadic function:
```
warning: second argument to 'va_arg' is of promotable type
      'short'; this va_arg has undefined behavior because arguments will be
      promoted to 'int' [-Wvarargs]
```

kren1 commented on test/Feature/VarArgByValReported.c:
```diff
@@ -0,0 +1,42 @@
+  va_list ap;
+  va_start(ap, x);
+  int i1 = va_arg(ap, int);
+  int i2 = va_arg(ap, int);
```
What about here? It seems less relevant.

kren1 commented:
Does this fix #247 too?

ccadar commented:
@MartinNowack thanks again for your help in debugging this!  Does my solution matches what we discussed?

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -1458,16 +1463,16 @@ void Executor::executeCall(ExecutionState &state,
 
-     // TODO: support "byval" parameter attribute
-     // TODO: support zeroext, signext, sret attributes
+    // TODO: support "byval" parameter attribute
```
This should be solved now ;)

ccadar commented:
@kren1 I have added a more comprehensive test.  @MartinNowack I removed that comment!
@kren1 this doesn't solve #247 (which is a related but different issue), but another PR is coming soon.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -1483,33 +1487,52 @@ void Executor::executeCall(ExecutionState &state,
           //
           // Alignment requirements for scalar types is the same as their size
-          if (argWidth > Expr::Int64) {
+          if (argWidths[k] > Expr::Int64 && !cs.isByValArgument(k)) {
```
I'm not sure if the `!cs.isByValArgument()` is correct - have a look at the email I sent you.

ccadar replied:
Sure, we can chat over email about this; let me know if you're understanding of the spec is different.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -1532,37 +1555,52 @@ void Executor::executeCall(ExecutionState &state,
-            Expr::Width argWidth = arguments[i]->getWidth();
-            if (argWidth > Expr::Int64) {
+          if (WordSize == Expr::Int64) {
+            if (argWidths[k] > Expr::Int64 && !cs.isByValArgument(k)) {
```
Same here.

ccadar commented:
@MartinNowack I cannot add `lli`crosschecking tests, as it is also buggy (as KLEE was before this patch).  I still think KLEE is the best interpreter for LLVM bitcode ;) 
```
$ lli --force-interpreter=true  VarArgByVal.bc 
Stack dump:
0.	Program arguments: lli --force-interpreter=true VarArgByVal.bc 
0  lli                      0x0000000107f16da8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  lli                      0x0000000107f17178 SignalHandler(int) + 180
2  libsystem_platform.dylib 0x00007fff7188a5fd _sigtramp + 29
3  lli                      0x0000000107be3c7f llvm::ExecutionEngine::getMangledName(llvm::GlobalValue const*) + 233
4  lli                      0x0000000107c07149 lle_X_memcpy(llvm::FunctionType*, llvm::ArrayRef<llvm::GenericValue>) + 90
5  lli                      0x0000000107c0548f llvm::Interpreter::callExternalFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 1119
6  lli                      0x0000000107bfc0d8 llvm::Interpreter::callFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 166
7  lli                      0x0000000107bfbce3 llvm::Interpreter::visitCallSite(llvm::CallSite) + 661
8  lli                      0x0000000107c0450a llvm::Interpreter::run() + 56
9  lli                      0x0000000107c086e3 llvm::Interpreter::runFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 57
10 lli                      0x0000000107be496a llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, char const* const*) + 926
11 lli                      0x00000001078ef10b main + 7971
12 libdyld.dylib            0x00007fff71691cc9 start + 1
Segmentation fault: 11
```
I plan to submit a bug report.

MartinNowack commented:
@ccadar I had a look at the PDF referenced from wikipedia (https://en.wikipedia.org/wiki/X86_calling_conventions#cite_note-AMD-22). And, I went through the LLVM code:

https://github.com/llvm/llvm-project/blob/d359f2096850c68b708bc25a7baca4282945949f/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L1811

In a nutshell, the pointer should be always updated according to alignment before reading from it independent of if the type is byval.

```
1)
auto alignment = kmodule->targetData->getStackAlignment();
if (argWidths[k] > alignment) {
// re-align
}

2) read from pointer

3) increment pointer by read typesize.
```

This should simplify the whole code and I guess the 32bit handling can be folded into 64bit handling.

ccadar commented:
Hi @MartinNowack thanks for the extra info!
I see your concerns, but:
```
struct mix {
  long long int first;
  int second;
  long long int third;
  long long int fourth;
};


void test(int x, ...) {
  va_list ap;
  va_start(ap, x);

  long long int ll = va_arg(ap, long long int);  
  fprintf(stderr, "ll: %lld\n", ll);
  
  struct mix m = va_arg(ap, struct mix);
  fprintf(stderr, "m: (%lld, %d, %lld, %lld)\n", m.first, m.second, m.third, m.fourth);
  // CHECK: m1: (7, 8, 9, 10)

  va_end(ap);
}

int main() {
  long long int l1 = 4;
  struct mix m1 = { 7, 8, 9, 10 };
  test(-1, l1, m1);
}
```
On x86_64: the `long long int` has size 8, the struct size 32.  I would expect the struct to be aligned at offset 16, but it is at offset 8.   So when I try to follow the spec (and your solution) things fail.  Am I misunderstanding the spec?

ccadar commented:
Hi again @MartinNowack , I think I got it!  I understand why the offset there is 8 and not 16, despite the spec saying it should be 16.  If you read the LLVM language spec at https://llvm.org/docs/LangRef.html, it says:

"The byval attribute also supports specifying an alignment with the align attribute. It indicates the alignment of the stack slot to form and the known alignment of the pointer specified to the call site. If the alignment is not specified, then the code generator makes a target-specific assumption."

And for the code above, LLVM generates this call:
`call void (i32, ...) @test(i32 -1, i64 %5, %struct.mix* byval(%struct.mix) align 8 %2)`

So what I know what is the missing part of the puzzle  now, I need to take into account the align attribute too.

ccadar commented:
Hi @MartinNowack  this is ready for re-review.  This should now follow the standard closely, as well as take into account the `align` attribute.

ccadar commented:
@MartinNowack ping?

MartinNowack commented:
@ccadar pong! :D

MartinNowack commented:
@ccadar I finally looked at this PR. I think the general concept is correct. But I see code duplication, which makes it harder to read.

Following suggestion: Instead of using the `argWidths` array to store widths of the object, it could be more efficient to remember the offsets into the svarg array in the first place.

And then store the arguments at the specific offsets. This avoids recalculating the offset and it would also make the storage independent of the used architecture.

ccadar commented:
@MartinNowack Thanks again for your comments.  I significantly refactored the patch to keep the offsets of variadic arguments rather than their sizes, which removed a lot of unnecessary code duplication (the code is now about half of what it was before).  I also merged a couple of commits.  Let me know if you have further suggestions for improvement.

MartinNowack commented:
@ccadar Excellent. I'm absolutely happy with those changes. The code is now much better to read. Let's merge this! Thanks for the effort.

ccadar commented:
Thank you as well for the excellent feedback.  It was essential both to understand some subtleties of variadic argument passing and to improve the code.  

So now one of our oldest bugs (almost 11 years old!) is fixed.   Now let me see if I can gather some strength to tackle byval arguments in a non-variadic argument context :)
